crates/kornia-algebra/src/vec.rs

```diff
@@ -0,0 +1,385 @@
+//! Macro to define a vector type.
+//!
+//! # Arguments
+//!
+//! * `name` - The name of the vector type.
+//! * `glam_type` - The underlying glam type.
+//! * `scalar` - The scalar type.
+//! * `array` - The array type.
+//! * `fields` - The fields of the vector.
+//!
+macro_rules! define_vector_type {
+    ($(#[$meta:meta])* $name:ident, $glam_type:ty, $scalar:ty, $array:ty, [$($field:ident),+]) => {
+        $(#[$meta])*
+        #[derive(Debug, Clone, Copy, PartialEq)]
+        pub struct $name {
+            $(pub $field: $scalar),+
+        }
+
+        impl $name {
+            /// Create a new vector from its components.
+            #[inline]
+            pub fn new($($field: $scalar),+) -> Self {
+                Self { $($field),+ }
+            }
+
+            /// Create a vector from an array.
+            #[inline]
+            pub fn from_array(arr: $array) -> Self {
+                let [$($field),+] = arr;
+                Self { $($field),+ }
+            }
+
+            /// Convert the vector to an array.
+            #[inline]
+            pub fn to_array(self) -> $array {
+                [$(self.$field),+]
+            }
+
+            /// Zero vector.
+            pub const ZERO: Self = Self {
+                $($field: 0.0 as $scalar),+
+            };
+
+            /// Euclidean length (magnitude) of the vector.
+            #[inline]
+            pub fn length(self) -> $scalar {
+                let v: $glam_type = self.into();
+                v.length()
+            }
+
+            /// Dot product between two vectors.
+            #[inline]
+            pub fn dot(self, rhs: Self) -> $scalar {
+                let a: $glam_type = self.into();
+                let b: $glam_type = rhs.into();
+                a.dot(b)
```

Comment:
Can’t we do self.0.dot(rhs.0) ? Wondering if all the into will create an overhead